plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx

```diff
@@ -193,69 +199,121 @@ export const getUsesReferencePreviewMessage = (
-		! isInRequiredLocation &&
-		! isProductContextSelected
-	) {
-		return ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER;
-	}
+			const { getEntityRecord, hasFinishedResolution } =
+				selectFunc( coreDataStore );
+			const selectorArgs = [ 'postType', 'product', productId ];
+			return {
+				product: getEntityRecord( ...selectorArgs ),
+				hasResolved: hasFinishedResolution(
+					'getEntityRecord',
+					selectorArgs
+				),
+			};
+		},
+		[ productId ]
+	);
+
+	const productCollectionUIStateInEditor = useMemo( () => {
+		const isInRequiredLocation = usesReference?.includes( location.type );
+		const isCollectionSelected = !! attributes.collection;
 
-	/**
-	 * Case 2: Preview mode - based on `usesReference` value
-	 */
-	if ( isInRequiredLocation ) {
 		/**
-		 * Block shouldn't be in preview mode when:
-		 * 1. Current location is archive and termId is available.
-		 * 2. Current location is product and productId is available.
-		 *
-		 * Because in these cases, we have required context on the editor side.
+		 * Case 1: Product context picker
 		 */
-		const isArchiveLocationWithTermId =
-			location.type === LocationType.Archive &&
-			( location.sourceData?.termId ?? null ) !== null;
-		const isProductLocationWithProductId =
-			location.type === LocationType.Product &&
-			( location.sourceData?.productId ?? null ) !== null;
+		const isProductContextRequired = usesReference?.includes( 'product' );
+		const isProductContextSelected =
+			( attributes.query?.productReference ?? null ) !== null;
+		if (
+			isCollectionSelected &&
+			isProductContextRequired &&
+			! isInRequiredLocation &&
+			! isProductContextSelected
+		) {
+			return ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER;
+		}
 
+		// Case 2: Deleted product reference
 		if (
-			! isArchiveLocationWithTermId &&
-			! isProductLocationWithProductId
+			isCollectionSelected &&
+			isProductContextRequired &&
+			! isInRequiredLocation &&
+			isProductContextSelected
 		) {
-			return ProductCollectionUIStatesInEditor.VALID_WITH_PREVIEW;
+			const isProductDeleted =
+				productId &&
+				( product === undefined || product?.status === 'trash' );
+			if ( isProductDeleted ) {
+				return ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE;
+			}
 		}
-	}
 
-	/**
-	 * Case 3: Collection chooser
-	 */
-	if ( ! hasInnerBlocks && ! isCollectionSelected ) {
-		return ProductCollectionUIStatesInEditor.COLLECTION_PICKER;
-	}
+		/**
+		 * Case 3: Preview mode - based on `usesReference` value
+		 */
+		if ( isInRequiredLocation ) {
+			/**
+			 * Block shouldn't be in preview mode when:
+			 * 1. Current location is archive and termId is available.
+			 * 2. Current location is product and productId is available.
+			 *
+			 * Because in these cases, we have required context on the editor side.
+			 */
+			const isArchiveLocationWithTermId =
+				location.type === LocationType.Archive &&
+				( location.sourceData?.termId ?? null ) !== null;
```

Comment:
nit: I think this is the same as
```suggestion
				location.sourceData?.termId != null;
```

Reply:
Right, But there is ESList rule because of which I can't use it 🤔  Therefore I went with current one.
![image](https://github.com/user-attachments/assets/d7825967-ab93-42ba-8c64-21ffbc99da2e)

Reply:
Ah, `eqeqeq`! 😅